contribute-to-tidb: update how to review a pr (#31)
zz-jason/tidb-dev-guide: contribute-to-tidb/review-a-pr.md

* We make sure to keep a high code quality in TiDB.
* We avoid situations where contributors and reviewers spend a lot of time refining a contribution that gets rejected later.

> ## 1. Things to know before the code review

Reviewer: Given that we don't use a checklist any more. It's better to remove the redundant `no.` prefix
```suggestion
## Things to know before the code review
```

Why is code review important?

> * Erase bugs early

Reviewer: ```suggestion
* Eliminate bugs early
```

* Make the project maintainable in the long run

> Some high level guidelines about code review in PingCAP (and mostly applies the TiDB community also):

Reviewer: ```suggestion
Some high level guidelines about code review in the TiDB community:
```

Yep. It could be guidelines in PingCAP. But they are also contributors in TiDB community and actually how TiDB community works. We don't have to emphasize "PingCAP" here.

Author: Okay... writing much more PRs than reviewing is common in our company, especially for the new employees.
For the community contributors, they don't have such strong feeling.

Reviewer: At least for this sentence, it emphasizes the importance of review itself. I don't feel an impact on "writing much more PRs than reviewing" among these words.

### Code review is more important than writing your own PR

> Modern software development is about cooperation, and nowadays software is no longer one single man’s piece of art. To speed up the development and make the processing scale, it is important to avoid blocking. If the members in an organization give higher priority for writing their own code, others waiting for code review get blocked, and there will be a lot of pending PRs for review. Eventually most of the developers get blocked, the whole team slows down.

Reviewer: Nice statement!

### It's the author's responsibility to get the PR merged

> We had a fierce debate on how to speed up the review process long long ago. What the role the reviewer and PR owner should play and what is each of their responsibility to get the PR merged faster.
> After the meeting, most of us agree with the conclusion that the PR owner is responsible for getting his PR merged. He is the one who most wanted to merge the PR into the repository. He is willing to lead that change. So he should try to do his best to have someone review his PR. The code reviewers should review the PR in time and give feedback, but it is not their responsibility to push the PR to be merged.

Reviewer: We can rephrase this section in "I" form instead of "he" for warming reading experience and avoid risk of sexism.

```suggestion
We had a fierce debate on how to speed up the review process long long ago. What the role the reviewer and PR owner should play and what is each of their responsibility to get the PR merged faster.

After the meeting, most of the participants agree with that the PR owner is responsible for getting PR merged.

If I am the PR author, I am the one who wants to merge the PR into the repository most. I'm willing to lead that change. So I'm supposed to do my best to have someone review the PR. The reviewers should review the PR in time and give feedback, but it is not their responsibility to push the PR to be merged.
```

> ### 2 LGTM for a PR to be accepted

Reviewer 2: It's changed to `Approve`, not `/LGTM`. Will you mention the method to approve the PR?

Reviewer: Yes. Please use Approval

```suggestion
### 2 Approvals for a PR to be accepted
```

> For most of the projects in PingCAP, we require 2 LGTM before the PR is accepted. There do have some exceptions, but this is the default setting. 2 LGTM ensures that even after the original PR author's resignation, there is someone familiar with the code and the project still keeps moving. And if that really happens, the code reviewer is the first person responsible for the code.

Reviewer: ```suggestion
For most of the repositories in the TiDB community, we require 2 approvals before the PR is accepted. A few repositories require different number of approvals, but 2 approvals is the default setting. 2 approvals ensures that even after the original PR author's resignation, there is someone familiar with the code and the project still keeps moving. And if the author really resigns, the code reviewer is the first person responsible for the code.
```

> ### Never give LGTM to to the code you don't understand

Reviewer: ```suggestion
### Never give approval to to the code you don't understand
```

A review should be thorough enough that the reviewer could explain the change at a reasonable level of detail to another developer. This ensures that the details of the code base are known to more than a single person. Remember: you are responsible for the code you reviewed, it is all about your reputation.

### No more than 300 lines for each PR

We encourage quick, small changes. 300 lines is basically normal people can comprehend within an hour. The reviewer has the right to ask the PR author to split the PR if it is too large. For the PR that is hard to split, the code reviewers can ask the author to have a meeting in person to explain the changes, until they fully understand the changes.

### One PR just does one thing

To make the PR easier to understand and get merged faster, this is also highly encouraged.

> For example, a bug fix should be in a separate PR, a refactoring should be in a separate PR. No matter how small it is, like updating the package version, or even one line changes to fix a typo, the change does exactly one thing and gets it right.

Reviewer: ```suggestion

For example, a bugfix should be in a separated PR, a refactoring should be in a separate PR. No matter how small it is, like updating the package version, or even one line changes to fix a typo, the change does exactly one thing and gets it right.
```

> Principles of the code review

Reviewer: ```suggestion
### Principles of the code review
```

* Technical facts and data overrules opinions and personal preferences

> * Software design is about trade-off and there is no perfect

Reviewer: ```suggestion
* Software design is about trade-off and there is no silver bullet
```

Everyone comes from different technical backgrounds with different knowledge. They have their own personal preferences. It is important that the code review is not based on biased opinions.

> A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, he should focus on what the bot can not do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.

Reviewer: ```suggestion
A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.
```

> Sometimes, making choices can be tricky. There is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, it means a huge maintenance burden. Should we accept or reject it? There is an emergency and someone has a PR to fix it, but that change is risky to introduce other kind of bugs. Should we accept or reject it?

Reviewer: ```suggestion
Sometimes, making choices can be tricky.

Said that there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, which means a huge maintenance burden. Shall we accept or reject it?

Said that there is a critical bug and a PR is read to fix it, but the change is risky to introduce other kind of bugs. Shall we accept or reject it?
```

> The answer is always “it depends”. Software design is more like a kind of art than technology. It is about aesthetics and your taste of the codeTradeoffs are always there and there is no perfect solution sometimes. Still, there is a general standard (from @shenli, maybe we can call it "better than before?"):

Reviewer: ```suggestion
The answer is always “it depends”. Software design is more like a kind of art than technology. It is about aesthetics and your taste of the code. Trade-offs are always there and there is no perfect solution sometimes. Still, there is a general standard (from [@shenli](https://github.com/shenli), maybe we can call it "better than before?"):
```

* It is making something better, although not a perfect solution.
* It can mitigate the pain of the user.
* There is no bug introduced by the change itself.

Then the code can be accepted.

Of course, code review should have some minimal requirements.

> For a bug fix PR, if there is no test case, ask the coder to add tests.

Reviewer: ```suggestion

For a bug fix PR, if it comes with no test case, the reviewer should ask the author to add tests.
```

> For a performance PR, if no benchmark result is provided, ask the coder to add a benchmark result.

Reviewer: ```suggestion

For a performance improvement PR, if no benchmark result is provided, the reviewer should ask the author to attach a benchmark result.
```

> ## 2. What to look for in a code review

Reviewer: ```suggestion
## What to look for in a code review
```

ditto

### Tests

Ask for unit, integration, or end-to-end tests as appropriate for the change. In general, tests should be added in the same PR as the production code.
Make sure that the tests in the PR are correct, sensible, and useful. Are they covering interesting cases? Are they readable? Does the PR lower overall test coverage? 
Think of ways this code could break. Will the tests actually fail when the code is broken? If the code changes beneath them, will they start producing false positives? Does each test make simple and useful assertions? Are the tests separated appropriately between different test methods?

Remember that tests are also code that has to be maintained. Don’t accept complexity in tests just because they aren’t part of the main binary.
Last but not least, keep an eye on how long the single unit takes to run. Does it slow the CI?

### Naming

> Did the developer pick good names for everything? A good name is long enough to fully communicate what the item is or does, without being so long that it becomes hard to read.

Reviewer: ```suggestion
Does the developer pick good names for everything? A good name is long enough to fully communicate what the item is or does, without being so long that it becomes hard to read.
```

### Comments

> Did the developer write clear comments in understandable English? Are all of the comments actually necessary? Usually comments are useful when they explain why some code exists, and should not be explaining what some code is doing. If the code isn’t clear enough to explain itself, then the code should be made simpler. There are some exceptions (regular expressions and complex algorithms often benefit greatly from comments that explain what they’re doing, for example) but mostly comments are for information that the code itself can’t possibly contain, like the reasoning behind a decision.
> It can also be helpful to look at comments that were there before this PR. Maybe there is a TODO that can be removed now, a comment advising against this change being made, etc.
> Note that comments are different from documentation of classes, modules, or functions, which should instead express the purpose of a piece of code, how it should be used, and how it behaves when used.

Reviewer: ```suggestion
* Does the developer write clear comments in understandable English?
* Are all of the comments actually necessary?

Usually comments are useful when they explain why the code exists, while not explain what the code is doing. If the code isn't clear enough to explain itself, then the code should be made simpler.

There are some exceptions; for example, regular expressions and complex algorithms often benefit greatly from comments that explain what they’re doing. But mostly comments are for information that the code itself can hardly contain, like the reason behind a decision.

It can also be helpful to look at comments existing before this PR. Maybe there is a `TODO` can be removed now, a comment advising against this change being made, etc.

Note that comments are different from documentation of classes, modules, or functions, which should instead express the purpose of a piece of code, how it should be used, and how it behaves when used.
```

### Style

> Make sure the PR follows the appropriate style guides. For Go and Rust, there are built in tools with the compiler toolchain.
> If you want to improve some style point that isn’t in the style guide, you can comment that you think it would improve the code readability but isn’t mandatory. Don’t block PRs from being submitted based only on personal style preferences.
> The author of the PR should not include major style changes combined with other changes. For example, if the author wants to reformat the whole file, have them send you just the reformatting as one PR, and then send another PR with their functional changes after that. This has been explained as “One PR just does one thing” in the previous section.

Reviewer: ```suggestion
Make sure the PR follows our [code style and quality guide](code-style-and-quality-guide.md). For Go and Rust, there are built in tools with the compiler toolchain.

If you want to improve style points outside the style guide, you can comment that you think it would improve the code readability but isn't mandatory. Don't block PRs from being submitted based only on personal style preferences.

The PR author should not include major style changes combined with other changes. For example, if a PR wants to reformat the whole file should not make functional changes. This has been explained as "One PR just does one thing" in the previous section.
```

### Consistency

What if the existing code is inconsistent with the style guide? The style guide is the absolute authority: if something is required by the style guide, the PR should follow the guidelines.

In some cases, the style guide makes recommendations rather than declaring requirements. In these cases, it’s a judgment call whether the new code should be consistent with the recommendations or the surrounding code. Bias towards following the style guide unless the local inconsistency would be too confusing

If no other rule applies, the author should maintain consistency with the existing code.

Either way, encourage the author to file a bug and add a TODO for cleaning up existing code.

### Documentation

> If a PR changes how users build, test, interact with, or release code, check to see that it also updates associated documentation, including READMEs, and any generated reference docs. If the PR deletes or deprecates code, consider whether the documentation should also be deleted. If documentation is missing, ask for it.
> Error Handling
> Be careful with error or exception handling. This is one of the most common places for bugs. Does the test cover the error code path? Does a panic make the whole process exit for server services? Does the code handle the resource releasing properly? Especially for Go programs involving concurrent goroutines, is there a goroutine leak or memory leak after the error?
> 
> Sometimes it is difficult for a real program to run to the error code path, and even more difficult to reproduce and debug if it really happens. Then it is necessary to mock the error code path. Tools like failpoint are available for that purpose.

Reviewer: ```suggestion
If a PR changes how users build, test, interact with, or release code, the reviewer are supposed to check whether it also updates related documentation, such as READMEs, and any generated reference docs. 

If a PR deletes or deprecates code, it should consider whether the documentation should also be deleted.

If documentation is missing, the reviewer should ask for it.

#### Error Handling

Be careful with error or exception handling. This is one of the most common places for bugs.

* Does the test cover the error code path?
* Does a panic make the whole process exit for server services?
* Does the code handle the resource releasing properly?
* Especially, for code including concurrent goroutines, is there a goroutine leak or memory leak after the error?

Sometimes it is difficult for a real program to run into the error code path, and even more difficult to reproduce and debug if it really happens. Then it is necessary to mock the error code path. Tools like [failpoint](https://github.com/pingcap/failpoint) are available for that purpose.
```

### Performance

Does the PR impose any performance impact? Sometimes a minor change on the hot path that seems unrelated may still affect performance. Is the algorithm in the PR suitable for that case?

> ## 3. How to write code review comments

Reviewer: ```suggestion
## How to write code review comments
```

ditto

### Be respectful to the reviewees

> Good reviewers are empathetic. Your code will always need to be reviewed. And you’ll always need to review your coworkers’ code. When you approach reviews as a learning process, everyone wins.

Reviewer: ```suggestion
Good reviewers are compassionate. Your code always need to be reviewed. And also you always need to review other contributors' code. Regard reviews as a learning process, and everyone wins.
```

### Ask questions rather than make statements

Wording is very important. Try to be constructive in your feedback, rather than critical. You can do this by asking questions, rather than making statements. And remember to give praise alongside your constructive feedback.

### Offer sincere praise

> Most reviewers focus only on what’s wrong with the code, but they should offer encouragement and appreciation for good practices, as well. It’s sometimes even more valuable, in terms of mentoring, to tell a developer what they did right than to tell them what they did wrong.

Reviewer: Definitely!

For example, imagine you’re reviewing for an author who struggles to write documentation, and you come across a clear, concise function comment. Let them know they nailed it. They will improve faster if you tell them when they got it right instead of just waiting to ding them when they screw up.

Summary

* Be kind to the coder, not to the code.
* Ask questions rather than make statements.
* Treat people who know less than you with respect, deference, and patience.
* Remember to praise when the code quality exceeds your expectation.
* It isn't necessarily wrong if the coder's solution is different from yours.
* Refer to the code style document when necessary.

> ## 4. Review with the @ti-chi-bot

Reviewer: ```suggestion
## Review with the @ti-chi-bot
```

ditto
